app/integration_action.go

```diff
@@ -31,90 +31,152 @@ import (
+		UserId: userId,
+		PostId: postId,
+	}
+
+	// See if the post exists in the DB, if so ignore the cookie.
+	// Start all queries here for parallel execution
 	pchan := a.Srv.Store.Post().GetSingle(postId)
 	cchan := a.Srv.Store.Channel().GetForPost(postId)
-
 	result := <-pchan
 	if result.Err != nil {
-		return "", result.Err
-	}
-	post := result.Data.(*model.Post)
+		if actionRequest.Cookie == "" {
+			return "", result.Err
+		}
 
-	result = <-cchan
-	if result.Err != nil {
-		return "", result.Err
-	}
-	channel := result.Data.(*model.Channel)
+		// Get action metadata from the cookie
+		c := model.PostActionCookie{}
+		cookie, err := a.decryptActionCookie(actionRequest.Cookie)
+		if err != nil {
+			return "", model.NewAppError("DoPostAction", "api.post.do_action.action_integration.app_error", nil, "err="+err.Error(), http.StatusBadRequest)
+		}
+		err = json.Unmarshal([]byte(cookie), &c)
+		if err != nil {
+			return "", model.NewAppError("DoPostAction", "api.post.do_action.action_integration.app_error", nil, "err="+err.Error(), http.StatusBadRequest)
+		}
+		if c.Integration == nil {
+			return "", model.NewAppError("DoPostAction", "api.post.do_action.action_integration.app_error", nil, "no Integration in action cookie", http.StatusBadRequest)
+		}
 
-	action := post.GetAction(actionId)
-	if action == nil || action.Integration == nil {
-		return "", model.NewAppError("DoPostAction", "api.post.do_action.action_id.app_error", nil, fmt.Sprintf("action=%v", action), http.StatusNotFound)
-	}
+		if postId != c.PostId {
+			return "", model.NewAppError("DoPostAction", "api.post.do_action.action_integration.app_error", nil, "postId doesn't match", http.StatusBadRequest)
+		}
+
+		upstreamRequest.ChannelId = c.ChannelId
+		upstreamRequest.TeamId = c.TeamId
+		upstreamRequest.Type = c.Type
+		upstreamRequest.Context = c.Integration.Context
+
+		if c.Type == model.POST_ACTION_TYPE_SELECT {
+			upstreamRequest.DataSource = c.DataSource
+			if upstreamRequest.Context == nil {
+				upstreamRequest.Context = map[string]interface{}{}
+			}
+			upstreamRequest.Context["selected_option"] = actionRequest.SelectedOption
+		}
+
+		retain = c.RetainProps
+		remove = c.RemoveProps
+		rootPostId = c.RootPostId
+		upstreamURL = c.Integration.URL
+	} else {
+		// Get action metadata from the database
+		post := result.Data.(*model.Post)
+
+		result = <-cchan
+		if result.Err != nil {
+			return "", result.Err
+		}
+		channel := result.Data.(*model.Channel)
+
+		action := post.GetAction(actionId)
+		if action == nil || action.Integration == nil {
+			return "", model.NewAppError("DoPostAction", "api.post.do_action.action_id.app_error", nil, fmt.Sprintf("action=%v", action), http.StatusNotFound)
+		}
+
+		upstreamRequest.ChannelId = post.ChannelId
+		upstreamRequest.TeamId = channel.TeamId
+		upstreamRequest.Type = action.Type
+		upstreamRequest.Context = action.Integration.Context
+
+		if action.Type == model.POST_ACTION_TYPE_SELECT {
```

Comment:
I copied the condition from prior code, but it seems unnecessary, 3/5 the fields should be initialized unconditionally, and the values just not be relevant for non-select types. Seems to me that the code here should avoid any knowledge of the action type semantics.

Reply:
+1

Reply:
Well, ended up not so simple... but seems correct now. Have to make sure Context gets created if nil, don't want to do it if there's nothing to save there, etc. Consolidated the logic for the 2 branches though (cookies and DB)